service/src/main/kotlin/fi/espoo/evaka/placement/PlacementTermination.kt

```diff
@@ -0,0 +1,203 @@
+    if (placement.startsAfter(terminationDate)) {
+        tx.cancelPlacement(placement.id)
+    } else {
+        tx.terminatePlacementFrom(
+            terminationRequestedDate,
+            placement.id,
+            terminationDate,
+            terminatedBy?.evakaUserId
+        )
+    }
+}
+
+fun terminateBilledDaycare(
+    tx: Database.Transaction,
+    user: AuthenticatedUser.Citizen,
+    terminationRequestedDate: LocalDate,
+    terminatablePlacementGroup: TerminatablePlacementGroup,
+    terminationDate: LocalDate,
+    childId: ChildId,
+    unitId: DaycareId
+) {
+    // additional placements after termination date are always cancelled
+    terminatablePlacementGroup.additionalPlacements
+        .filter { it.endDate.isAfter(terminationDate) }
+        .forEach { cancelOrTerminatePlacement(tx, terminationRequestedDate, it, terminationDate, user) }
+
+    val preschoolOrPreparatoryWithDaycare = terminatablePlacementGroup.placements
+        .filter {
+            (it.type == PRESCHOOL_DAYCARE || it.type == PREPARATORY_DAYCARE) && it.endDate.isAfter(terminationDate)
+        }
+    // placements which have already been updated on the earlier iteration
+    val placementsToSkip = mutableListOf<ChildPlacement>()
+    for (placement in preschoolOrPreparatoryWithDaycare) {
+        if (placementsToSkip.contains(placement)) {
+            continue
+        }
+        // It is possible that we already have an adjacent placement of the same type, e.g. when the billed DAYCARE part
+        // of a PRESCHOOL_DAYCARE placement has been terminated before. Detect those and prevent duplicates.
+        val adjacentPlacement = terminatablePlacementGroup.placements.find {
+            it.type == when (placement.type) {
+                PRESCHOOL_DAYCARE -> PRESCHOOL
+                PREPARATORY_DAYCARE -> PREPARATORY
+                else -> throw IllegalStateException("Should not be handling any other placement types here")
+            } && it.startDate == placement.endDate.plusDays(1)
+        }?.also { placementsToSkip.add(it) }
+
+        // If placement is in the future
+        //  - has an adjacent placement => cancel placement & adjust adjacent start
+        //  - no adjacent => change PRESCHOOL_DAYCARE to PRESCHOOL
+        // If placement is ongoing
+        // - has an adjacent placement => change the end date to termination date & adjust adjacent start
+        // - no adjacent => change the end date and create adjacent
+
+        if (placement.startDate.isAfter(terminationDate)) {
+            if (adjacentPlacement != null) {
+                // given an adjacent placement, do not create a duplicate PRESCHOOL or PREPARATORY placement, just update the dates
+                tx.cancelPlacement(placement.id)
+                tx.updatePlacementStartDate(adjacentPlacement.id, placement.startDate)
+            } else {
+                // convert all future PRESCHOOL_DAYCARE and PREPARATORY_DAYCARE placements to PRESCHOOL and PREPARATORY
```

Comment:
I think talking about all future placements is a bit confusing here since only the currently handled placement is affected.